src/Altinn.ApiClients.Maskinporten/Service/MaskinportenService.cs

```diff
@@ -0,0 +1,227 @@
+        {
+            X509SecurityKey securityKey = new X509SecurityKey(cert);
+            JwtHeader header = new JwtHeader(new SigningCredentials(securityKey, SecurityAlgorithms.RsaSha256))
+            {
+                { "x5c", new List<string>() { Convert.ToBase64String(cert.GetRawCertData()) } }
+            };
+            header.Remove("typ");
+            header.Remove("kid");
+            return header;
+        }
+
+        private FormUrlEncodedContent GetUrlEncodedContent(string assertion)
+        {
+            FormUrlEncodedContent formContent = new FormUrlEncodedContent(new List<KeyValuePair<string, string>>
+            {
+                new KeyValuePair<string, string>("grant_type", "urn:ietf:params:oauth:grant-type:jwt-bearer"),
+                new KeyValuePair<string, string>("assertion", assertion),
+            });
+
+            return formContent;
+        }
+
+        public async Task<TokenResponse> PostToken(FormUrlEncodedContent bearer)
+        {
+            TokenResponse token = null; ;
+
+            HttpRequestMessage requestMessage = new HttpRequestMessage()
+            {
+                Method = HttpMethod.Post,
+                RequestUri = new Uri(GetTokenEndpoint()),
+                Content = bearer
+            };
+
+            HttpResponseMessage response = await _client.SendAsync(requestMessage);
+
+            if (response.IsSuccessStatusCode)
+            {
+                string content = await response.Content.ReadAsStringAsync();
+                token = System.Text.Json.JsonSerializer.Deserialize<TokenResponse>(content);
+                return token;
+            }
+            else
+            {
+                string error = await response.Content.ReadAsStringAsync();
+            }
+
+            return null;
+        }
+
+        private string GetAssertionAud()
+        {
+            if (_maskinportenConfig.Environment.Equals("prod"))
+            {
+                return _maskinportenConfig.JwtAssertionAudienceProd;
+            }
+            else if (_maskinportenConfig.Environment.Equals("ver1"))
+            {
+                return _maskinportenConfig.JwtAssertionAudienceVer1;
+            }
+            else if (_maskinportenConfig.Environment.Equals("ver2"))
+            {
+                return _maskinportenConfig.JwtAssertionAudienceVer2;
+            }
+
+            return null;
+        }
+
+        private string GetTokenEndpoint()
+        {
+           if(_maskinportenConfig.Environment.Equals("prod"))
+            {
+                return _maskinportenConfig.TokenEndpointProd;
+            }
+            else if (_maskinportenConfig.Environment.Equals("ver1"))
+            {
+                return _maskinportenConfig.TokenEndpointVer1;
+            }
+            else if (_maskinportenConfig.Environment.Equals("ver2"))
+            {
+                return _maskinportenConfig.TokenEndpointVer2;
+            }
+
+           return null;
```

Comment:
Perhaps check if a "TokenEndpoint" setting has been suppled (allowing preset values to be overridden)? If not, we should probably just throw.

Reply:
I am not sure I understand. Since it is provided as standard it would mean they have overwritten with blank?

Reply:
Not a big deal, but the idea was that we could make it possible for a user to supply `TokenEndpoint` and `JwtAssertionAudience` in settings as an alternative (override) to supplying `Environment`, which may be useful for testing. 

In any case, I don't think returning null here makes any sense here; it will just result in a not-immediately-obvious NullReferenceException thrown elsewhere. It should instead just `throw new ArgumentException('Invalid environment supplied, valid values are: ver1, ver2, prod')`